### DISCUSSION TITLE
In the strongswan+openssl scenario, error relocating libstrongswan-openssl.so

### DISCUSSION THREAD

**Discussion by kamen-geats** [strongswan/strongswan#1715] [Q&A]
Due to some historical reasons, another _libcrypto.so_ (generated through mbed TLS) exists in the environment, We can only try to rename the _libcrypto.so_ compiled by openssl(eg: libopensslcrypto.so). The _libstrongswan-openssl.so_  fails to be loaded, the following is the error log:

> 00[LIB] plugin 'openssl' failed to load: Error relocating /system/lib/ipsec/plugins/libstrongswan-openssl.so: BN_div: symbol not found

We temporarily deleted the _libcrypto.so_(generated through mbed TLS),  rename libopensslcrypto.so to libcrypto.so,  the error can be solved.

Can we use the configuration file to specify the name of the linked dynamic library, and if so, I think it solves the current problem. Thanks!

**Comment by tobiasbrunner** [comment]
> Can we use the configuration file to specify the name of the linked dynamic library, and if so, I think it solves the current problem.

You could try setting `openssl_lib` when running the configure script:

     ./configure --enable-openssl ... openssl_lib=opensslcrypto

Or if there are two `libcrypto.so` files in different locations, specify `LDFLAGS` with the path to the library you want to use when running the script.

**Reply by kamen-geats** [reply]
We tried to set `openssl_lib` when running the configure script, but there's still an error here:

> plugin 'openssl' failed to load: Error loading shared library libcrypto.so.1.0.0: No such file or directory (needed by /system/lib/ipsec/plugins/libstrongswan-openssl.so)
> connecting to 'unix:///var/run/charon.vici' failed: No such file or directory                                                       
> Error: connecting to 'default' URI failed: No such file or directory

I renamed libopensslcrypto.so to libcrypto.so.1.0.0, however, the charon.vici error still exists. The parameters of the configure script are as follows, is there anything missing from our configuration?

>      ./configure --host=$(ARCH)-linux \
> 	$(TARGET_ATP_ENVS) CFLAGS="-I$(STAGING_DIR)/usr/include $(SECURE_CFLAGS) -D__UAPI_DEF_ETHHDR" \
> 	LDFLAGS="$(SECURE_LDFLAGS) $(PIE_SECURE_LDFLAGS) -s" CPPFLAGS="-I$(STAGING_DIR)/usr/include" \
> 	LIBS="-L$(TARGET_DIR)$(TARGET_SYSTEM)/lib -lgplutil" LT_SYS_LIBRARY_PATH="system/lib/ipsec/openssl"\
> 	SYSROOT="$(STAGING_DIR)"  \
> 	--includedir=$(STAGING_DIR)/usr/include \
> 	--oldincludedir=$(STAGING_DIR)/usr/include \
> 	--enable-openssl \
> 	openssl_lib=opensslcrypto \
> 	--enable-counters \
> 	--disable-gmp \
> 	--disable-stroke \
> 	--disable-pki \
> 	--disable-scepclient \
> 	--prefix=/system \
> 	--libexecdir=/system/bin \
> 	--sysconfdir=/var/vpn \
> 	--with-lib-prefix=$(TARGET_DIR)$(TARGET_SYSTEM)/lib \
> 	--with-sysroot=$(STAGING_DIR) \
> 	--with-user=strongswan \
> 	--with-group=strongswan \
> 	--with-capabilities=native \

**Reply by tobiasbrunner** [reply]
Did you run `make clean` before or after you changed the configure options? Also, `LT_SYS_LIBRARY_PATH` seems to be set to a relative path, you might want to add a `/` there.

**Reply by kamen-geats** [reply]
Yes, before compiling, we delete the previous strongswan source code and copy a new one.
The value of `LT_SYS_LIBRARY_PATH` is an absolute path, the absolute path on the board starts with `system`.

However, I still think `libcrypto.so.1.0.0` in the log is incorrect. Can I change the name of the shared library of load(eg:libcrypto.so.1.0.0 -> libopensslcrypto.so)? I'm trying to change the name of `libcrypto.so.1.0.0` temporarily here, however, the following error message is still displayed:

> connecting to 'unix:///var/run/charon.vici' failed: No such file or directory
> Error: connecting to 'default' URI failed: No such file or directory
> 

I didn't encounter this error before setting `openssl_lib`. Is this error related to setting` openssl_lib`?Thank you in advance!

**Reply by tobiasbrunner** [reply]
> The value of `LT_SYS_LIBRARY_PATH` is an absolute path,

It wasn't in the output you posted above.

> However, I still think `libcrypto.so.1.0.0` in the log is incorrect.

Could it be that you just renamed a symlink that points to that file?

> I didn't encounter this error before setting `openssl_lib`.

That's unrelated. It's normal if you run swanctl while the daemon is not running.

**Reply by kamen-geats** [reply]
Hi, we've revised the previous questions, now there's one last question left:

> 00[LIB] plugin 'hmac': loaded successfully
> 00[LIB] plugin 'pkcs1': loaded successfully
> 00[LIB] plugin 'curve25519': loaded successfully
> 00[LIB] plugin 'openssl' failed to load: Error loading shared library **libcrypto.so.1.0.0**: No such file or directory (needed by /system/lib/ipsec/plugins/libstrongswan-openssl.so)
> 00[LIB] plugin 'kernel-netlink': loaded successfully

Which of the following is the logic restriction of `libcrypto.so.1.0.0` in the code? How can we change the name of the `libcrypto.so.1.0.0` here? In the code, libcrypto.so.1.0.0 is contained only in the `data.sql`.

> line:708
> INSERT INTO files (				/*  1 */
>   name, dir
> ) VALUES (
>  'libcrypto.so.1.0.0', 5
> );

We tried to modify it, but it didn't work.

**Reply by tobiasbrunner** [reply]
That has absolutely nothing to do with it. That SQL data is related to [TNC](https://docs.strongswan.org/docs/5.9/tnc/tnc.html).

As I suggested before, the error could be due to an issue with symlinks.

**Comment by kamen-geats** [comment]
> 00[LIB] plugin 'openssl' failed to load: Error loading shared library libcrypto.so.1.0.0: No such file or directory (needed by /system/lib/ipsec/plugins/libstrongswan-openssl.so)

It's not this mistake that's causing the problem, this is because` libcrypto.so.1.0.0`(compiled by mbedtls) already exists in the environment,  we need `strongswan` to link to another `libopensslcrypto.so`(compiled by openssl). However, the configuration does not take effect. The following is the error log:

> 00[LIB] plugin 'openssl' failed to load: Error relocating /system/lib/ipsec/plugins/libstrongswan-openssl.so: BN_div: symbol not found

**Reply by kamen-geats** [reply]
In other words, does strongswan support **static links** to openssl, and if so, is there a description of how to do it?

**Reply by tobiasbrunner** [reply]
> we need `strongswan` to link to another `libopensslcrypto.so`(compiled by openssl).

That should be the case if you set `openssl_lib`.  But if you just renamed `libcrypto.so` to `libopensslcrypto.so` that won't help if that file is a symlink that points to `libcrypto.so.1.0.0`. Also consider `LD_LIBRARY_PATH` (see `man ld`).

> The following is the error log:

I guess that happens if mbedtls's libcrypto is loaded, which apparently doesn't provide `BN_div()` (it's possible that this is a macros in the mbedtls headers, which isn't replaced if you compile with the OpenSSL headers).

> In other words, does strongswan support **static links** to openssl, and if so, is there a description of how to do it?

I think it's possible to statically link `libcrypto.a` to the _openssl_ plugin. But you'd have to change the `Makefile.am` file of the _openssl_ plugin. Something like this might work:

```diff
diff --git a/src/libstrongswan/plugins/openssl/Makefile.am b/src/libstrongswan/plugins/openssl/Makefile.am
index 979bd8886138..abfd34c776b3 100644
--- a/src/libstrongswan/plugins/openssl/Makefile.am
+++ b/src/libstrongswan/plugins/openssl/Makefile.am
@@ -42,4 +42,4 @@ libstrongswan_openssl_la_SOURCES = \
 	openssl_xof.c openssl_xof.h
 
 libstrongswan_openssl_la_LDFLAGS = -module -avoid-version
-libstrongswan_openssl_la_LIBADD  = $(OPENSSL_LIB)
+libstrongswan_openssl_la_LIBADD  = /path/to/libcrypto.a
```
Adding `-Wl,-Bstatic $(OPENSSL_LIB) -Wl,-Bdynamic` to `libstrongswan_openssl_la_LDFLAGS` sadly doesn't work because libtool reorders the options (to something like this: ` $(OPENSSL_LIB) -Wl,-Bstatic -Wl,-Bdynamic`) and still links dynamically.
